[Teo Rayon] iP

NoyaRoeT opened the pull request:
# Duke
> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Duke remembers things you need to do, so you don't have to. It's
- text-based
- easy to use
- ~~SUPER~~  _**MEGA**_ FAST to use

All you need to do is,
1. download it
2. double-click it
3. add your tasks
4. let it manage tasks for you :grin:

It's also **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

If you're a Java programmer, you may see some familiar code. Here's the `showMessage` method:
```java
public void showMessage(String msg) {
        String wrapTop = "__________________________\n";
        String wrapBottom = "\n__________________________";
        System.out.println(wrapTop + msg.toString() + wrapBottom);
}
```

Commit:
Add Gradle support

NoyaRoeT committed:
Implemented Level-1

NoyaRoeT committed:
Implemented Level-2

NoyaRoeT committed:
Implemented Level-3

NoyaRoeT committed:
Implemented Level-4

NoyaRoeT committed:
Implemented A-TextUiTesting

NoyaRoeT committed:
Implemented Level-5: Handling Errors

NoyaRoeT committed:
Level-6

NoyaRoeT committed:
Add save to and read from file functionality

NoyaRoeT committed:
Add date and time parsing + formatting

NoyaRoeT committed:
Merge branch 'branch-level-7'

NoyaRoeT committed:
Merge branch-level-8 to master

NoyaRoeT committed:
Remove txt files

NoyaRoeT committed:
Add storage class

NoyaRoeT committed:
Add TaskList class

NoyaRoeT committed:
Add parser class

NoyaRoeT committed:
Add Ui class

NoyaRoeT committed:
Move classes to package main.java.duke

NoyaRoeT committed:
Move classes to duke package

NoyaRoeT committed:
Merge remote-tracking branch 'origin/add-gradle-support'

NoyaRoeT committed:
Add gradle build support

NoyaRoeT committed:
Fix mainClassName in build.gradle

NoyaRoeT committed:
Add JUnit tests for DateTimeParser and Deadline

NoyaRoeT committed:
Update error messages

NoyaRoeT committed:
Add JavaDoc comments to public methods

NoyaRoeT committed:
Update classes to adhere to coding standards

NoyaRoeT committed:
Add find by keyword functionality

NoyaRoeT committed:
Merge branch 'branch-A-JavaDoc'

NoyaRoeT committed:
Merge branch 'branch-A-CodingStandard'

NoyaRoeT committed:
Merge branch 'branch-Level-9'

NoyaRoeT committed:
Add header comments for public classes

NoyaRoeT committed:
Add GUI for Duke

chen-jerry-junior commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,39 @@
+    public void saveTaskList() {
+        try {
+            storage.save(taskList.toStorageString());
+        } catch(IOException e) {
```
I think there should be a space after "catch" due to Checkstyle maybe?

chen-jerry-junior commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,171 @@
+        Scanner stringStream = new Scanner(commandString);
+        String command = stringStream.next();
+
+        if (command.equalsIgnoreCase("bye")) {
```
I like your idea of using eauilsIgnoreCase()

chen-jerry-junior commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,171 @@
+        }
+    }
+
+    private static String handleList(TaskList taskList) {
```
For the following handle~ methods, maybe you can try to encapsulated them into a new packet called Command and create different types of commands there.

chen-jerry-junior commented on src/main/java/duke/DateTimeParser.java:
```diff
@@ -0,0 +1,127 @@
+ * Contains functionality for parsing date and time input, as well as formatting
+ * LocalDate and LocalDateTime objects.
+ */
+public class DateTimeParser {
```
I like your idea of creating a new class for parsing LocalDateTime. Maybe it will be better if the parser and datetimeparser are under the same packet?

chen-jerry-junior reviewed (commented):
LGTM, almost. Just a few nits to fix.

nramapurath commented on src/main/java/duke/DateTimeParser.java:
```diff
@@ -0,0 +1,127 @@
+ */
+public class DateTimeParser {
+
+    private static DateTimeFormatter parser = new DateTimeFormatterBuilder()
```
I was impressed by your DateTimeParser (did not know of these methods I could use). Your use of appendOptionals to give the user more options to input the date has also inspired me to do better! However, Maybe you could have added a short comment here to explain the parser member as there does seem to be some implicit logic involved?

nramapurath commented on src/main/java/duke/DateTimeParser.java:
```diff
@@ -0,0 +1,127 @@
+     * Returns a LocalDateTime object from the given string containing date and time.
+     * The format for the date and time string is "dd-mm-yyyy hh:ss". The time may be omitted.
+     *
+     * @param dateTimeString a date and time string with the format "dd-mm-yyyy hh:ss"
```
Very minor nit here but according to the [coding standard for Javadocs](https://se-education.org/guides/conventions/java/intermediate.html#comments), each parameter description should be punctuated. Perhaps you could add in a period at the end and capitalize the first letter after the parameter (`dateTimeString A date and time string with the format "dd-mm-yyyy hh:ss".`)

nramapurath commented on src/main/java/duke/DateTimeParser.java:
```diff
@@ -0,0 +1,127 @@
+            return "th";
+        }
+        switch (day % 10) {
+        case 1: return "st";
```
Minor nit here but the [coding standard](https://se-education.org/guides/conventions/java/intermediate.html#layout) does seem to require an explicit `// Fallthrough` comment whenever there is a switch case without a break statement.

nramapurath reviewed (commented):
Overall, code looks very well done! I can see that a a good amount of planning went into it. I only had a couple nits regarding punctuation in the Javadocs overall and a couple missing `// Fallthrough` comments in switch cases without break statements.

NoyaRoeT committed:
Merge branch 'branch-Level-10'

NoyaRoeT committed:
Add abstract Command class and subclasses representing each command

NoyaRoeT committed:
Add DukeResponse class and fix bye command bug

NoyaRoeT committed:
Replace Ui class with MessageGenerator class

NoyaRoeT committed:
Add assertions to verify inputs generated by code.

Methods that take in internally generated input assumes that the inputs
are correct.

However, this depends on correctness in code that generate these
inputs. Using assertions to document these assumptions allow us to
verify program correctness.

Let's add assertions to methods that take in inputs generated
internally.

Using assertions is preferable over exceptions, since these internally
generated inputs should not ever be invalid and should not be recovered
from.

NoyaRoeT committed:
Shorten too long methods, group similar methods.

There are many long methods in Parser for handling user commands.
Notably the handleDeadline and handleEvent commands. There are also
methods with very similar code except for one or two lines.

Long methods or methods with almost the same code are a result of poor abstraction. Proper abstraction can reduce code duplication and complexity by separating responsibility.

Let's move methods with very similar bodies such as handleMark,
handleUnmark and handleDelete into a single method handleTaskIdxCommand.
Let's also make Parser only responsible for parsing user input into
apporpriate strings. The commands themselves will transform these
strings into the appropriate data types.

This way, we reduce code duplication for commands that require a task
index. We also clearly define the responbility of the Parser class.

NoyaRoeT committed:
Merge pull request #1 from NoyaRoeT/branch-A-Assertions

Add assertions to verify inputs generated by code.

NoyaRoeT committed:
Merge master into branch-A-CodeQuality

NoyaRoeT committed:
Merge master into branch-A-CodeQuality

NoyaRoeT committed:
Merge pull request #3 from NoyaRoeT/branch-A-CodeQuality

Shorten too long methods, group similar methods.

NoyaRoeT committed:
Add ability to update field of a task

NoyaRoeT committed:
Merge branch 'branch-C-Update'

NoyaRoeT committed:
Redesign GUI

NoyaRoeT committed:
Resize text for DialogBox

NoyaRoeT committed:
Add Ui image to docs/

NoyaRoeT committed:
Rename Ui.PNG to Ui.png

NoyaRoeT committed:
Update README.md

NoyaRoeT committed:
Add files via upload

NoyaRoeT committed:
Update README.md

NoyaRoeT committed:
Fix image paths

NoyaRoeT committed:
Change image paths

NoyaRoeT committed:
Update image paths

NoyaRoeT committed:
Increase spacing between features

NoyaRoeT committed:
Add more test cases for DateTimeParser

NoyaRoeT committed:
Merge branch 'master' of https://github.com/NoyaRoeT/ip

NoyaRoeT committed:
Fix checkstyle violations

NoyaRoeT committed:
Fix bug with task numbers when using find
